ingest/indexer/service/indexer_streaming_service_test.go

```diff
@@ -0,0 +1,423 @@
+			}
+
+			// Initialize an empty pool tracker
+			emptyPoolTracker := pooltracker.NewMemory()
+
+			// Initialize a mock pool extractor
+			poolExtractorMock := &sqsmocks.PoolsExtractorMock{
+				BlockPools: commondomain.BlockPools{
+					ConcentratedPools: concentratedPools,
+					CFMMPools:         cfmmPools,
+					CosmWasmPools:     cosmWasmPools,
+				},
+			}
+
+			// Initialize a block process strategy manager
+			blockProcessStrategyManager := commondomain.NewBlockProcessStrategyManager()
+
+			// Initialize a concentrated pool with spread factor
+			sf, _ := math.LegacyNewDecFromStr(defaultSpreadFactor)
+			s.CreateConcentratedPoolsAndFullRangePositionWithSpreadFactor(defaultPoolDenoms, []math.LegacyDec{sf})
+			pools, err := s.App.ConcentratedLiquidityKeeper.GetPools(s.Ctx)
+			s.Require().NoError(err)
+			pool := pools[1]
+			poolID := pool.GetId()
+
+			// Initialize keepers
+			keepers := indexerdomain.Keepers{
+				PoolManagerKeeper: s.App.PoolManagerKeeper,
+			}
+
+			// Initialize tx decoder and logger
+			txDecoder := s.App.GetTxConfig().TxDecoder()
+			logger := s.App.Logger()
+
+			// Initialize a mock publisher
+			publisherMock := &indexermocks.PublisherMock{}
+
+			// Initialize an indexer streaming service
+			indexerStreamingService := indexerservice.New(
+				emptyWriteListeners,
+				blockProcessStrategyManager,
+				publisherMock,
+				poolExtractorMock,
+				emptyPoolTracker,
+				keepers,
+				txDecoder,
+				logger)
+
+			// Create the event based on the test cases attributes
+			event := abcitypes.Event{
+				Type: tc.eventType,
+				Attributes: func() []abcitypes.EventAttribute {
+					if tc.havePoolIDAttribute {
+						return []abcitypes.EventAttribute{
+							{
+								Key: concentratedliquiditytypes.AttributeKeyPoolId,
+								Value: func() string {
+									if tc.useCLPoolID {
+										return strconv.Itoa(int(poolID))
+									} else {
+										return strconv.Itoa(rand.Intn(1000) + 1000)
+									}
+								}(),
+								Index: false,
+							},
+						}
+					} else {
+						return []abcitypes.EventAttribute{}
+					}
+				}(),
+			}
+
+			// Pass the event to the AddTokenLiquidity method
+			err = indexerStreamingService.AddTokenLiquidity(s.Ctx, &event)
+			s.Require().Equal(tc.expectedError, err != nil)
+
+			// Assert the "liquidity_{denom}"" event attribute
+			denoms := pool.GetPoolDenoms(s.Ctx)
+			liquidityKey0 := liquidityAttributePrefix + denoms[0]
+			liquidityKey1 := liquidityAttributePrefix + denoms[1]
+			var foundKey0, foundKey1 bool
+			for _, attribute := range event.Attributes {
+				if attribute.Key == liquidityKey0 {
+					foundKey0 = true
+					continue
+				}
+				if attribute.Key == liquidityKey1 {
+					foundKey1 = true
+					continue
+				}
+			}
+			if tc.expectedLiquidity {
+				s.Require().True(foundKey0)
+				s.Require().True(foundKey1)
+			} else {
+				s.Require().False(foundKey0)
+				s.Require().False(foundKey1)
+			}
+
+		})
```

Comment:
I wonder if we could encapsulate some of the verbose boilerplate into test helper methods so that the core test logic is more simple and readable.

I currently feel like there is a lot going on in the test

Reply:
Good point. Done by externalize such block of the code into a function named `checkIfLiquidityAttributeExists`